### PR TITLE
Refactor auth-related functions + CSRF token and cookie matching

### PR DESCRIPTION
**Pull request by isoos**
- https://github.com/dart-lang/pub-dartlang-dart/issues/1642
- the CSRF token matching is a bit "hanging in the air", but will be used later

### CODE REVIEW

**Comment by sortie** [comment]
@jonasfj Can you review this one?

File: app/lib/frontend/auth.dart [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        registerLoggedInUser(email);
+      }
+    }
+  }
+}
+
+/// Whether the request has a valid pair of CSRF tokens in the cookie and the
+/// request headers. This is the validation part of the cookie-to-header pattern
+/// https://en.wikipedia.org/wiki/Cross-site_request_forgery#Cookie-to-header_token
+bool isCsrfCookieAndHeaderValid(shelf.Request request) {
```
Comment (jonasfj) [inline app/lib/frontend/auth.dart]: Why do we need this? Are we planning to use cookies as a separate authentication mechanism?

We currently use an oauth2 `accessToken` for authorization, can't we keep doing that, and just use a different flow when signing in on the pub site:
https://developers.google.com/identity/protocols/OAuth2UserAgent

Then we'll have an `accessToken` in javascript, and whatever ajax application we build can talk to the same APIs as the `pub` client. Hence, we won't need to implement the APIs twice, we don't need to worry about CSRF, but we still have to worry about XSS :)

I assume this in preparation of being able to login to the pub site.

Comment (isoos) [reply app/lib/frontend/auth.dart]: I'd rather assume we will use a mix of APIs, some overlapping with the `pub` client.

I believe `CSRF` will be an issue, no matter what kind of access token we are using: the auth cookie is sent on every request, and if an attacker can create a forged request, we won't be able to distinguish it without some kind of extra check. It seems to me that they recommend even the authorization flow to be [protected by CSRF tokens](https://security.stackexchange.com/questions/20187/oauth2-cross-site-request-forgery-and-state-parameter).

**Comment by isoos** [comment]
obsolete for now

**Pull request closed without merging**